content/articles/how-to-debug-a-nodejs-application-running-in-a-docker-container/index.md

```diff
@@ -0,0 +1,489 @@
+
+![packagejson](/engineering-education/how-to-debug-a-nodejs-application-running-in-a-docker-container/packagejson.PNG)
+
+### Creating the to-do application
+Next, we will build our to-do application using [Express](https://expressjs.com/): a fast, popular and lightweight Node.js web framework.
+
+Express makes the development of web applications easier.
+
+#### Step 1: Installing project's dependencies
+We will install `express` and a few other prerequisites needed by the application by running the command below:
+
+```bash
+$ npm install express body-parser cookie-session ejs --save
+```
+
+#### Step 2: Coding the app.js file
+Then we will create a file named `app.js` in the `TodoApp` directory with the code below:
+
+```javascript
+const express = require('express')
+const app = express()
+const bodyParser = require('body-parser')
+const session = require('cookie-session')
+const urlencodedParser = bodyParser.urlencoded({ extended: false })
+const port = 3000
+
+app.use(session({ secret: process.env.SECRET }))
+  .use(function (req, res, next) {
+    if (typeof (req.session.todolist) == 'undefined') {
+      req.session.todolist = []
+  }
+    next()
+  })
+
+  .get ('/todo', function (req, res) {
+    res.render('todo.ejs', { todolist: req.session.todolist })
+  })
+
+  .post ('/todo/add/', urlencodedParser, function (req, res) {
+    if (req.body.newtodo != '') {
+      req.session.todolist.push(req.body.newtodo)
+    }
+    res.redirect('/todo')
+  })
+
+  .get ('/todo/delete/:id', function (req, res) {
+    if (req.params.id != '') {
+      req.session.todolist.splice(req.params.id, 1)
+    }
+    res.redirect('/todo')
+  })
+
+  .use (function (req, res, next) {
+    res.redirect('/todo')
+  })
+
+
+  .listen(port, () => console.log(`MyTodo app is listening on port ${port}!`))
+```
+
+#### Step 3: Creating the todo.ejs file
+Next we will create a file `./views/todo.ejs` and paste the following code in it:
+
+```html
+<!DOCTYPE html>
+
+<html>
+    <head>
+        <title>My todolist</title>
+        <style>
+            a {text-decoration: none; color: black;}
+        </style>
+    </head>
+
+    <body>
+        <h1>My todolist</h1>
+
+        <ul>
+        <% todolist.forEach(function(todo, index) { %>
+            <li><a href="/todo/delete/<%= index %>">✘</a> <%= todo %></li>
+        <% }); %>
+        </ul>
+
+        <form action="/todo/add/" method="post">
+            <p>
+                <label for="newtodo">What should I do?</label>
+                <input type="text" name="newtodo" id="newtodo" autofocus />
+                <input type="submit" />
+            </p>
+        </form>
+    </body>
+</html>
+```
+
+#### Step 4: Starting the web server
+Start the web server by executing the command below:
+
+```bash
+$ export SECRET=bestkeptsecret; node app.js
+```
```

Comment:
@verah-tech There are no bugs in this code. Please do something 😅

Reply:
@geoffrey45 The bug was not on this specific line, it was on the line I placed a breakpoint, I can remove the keyword export on this mentioned line if it works well on your end.
Thank you

Reply:
I placed the commet here because this is where you run the app. Running the app allows adding todos without any bug. Also the fix at the bottom of this document is already in the initial code.

Reply:
@geoffrey45 I mistakenly copied the final code in the initial code from my testing environment, which was supposed to have a bug. I have fixed the issue, kindly review.
Thank you